pallets/settlement/src/lib.rs

```diff
@@ -1295,13 +1313,69 @@ impl<T: Config> Module<T> {
         Ok(())
     }
 
+    /// Returns `Ok` if all affirmations have been received. Otherwise, returns an error.
+    /// This call also removes all elements from `UserAffirmations`, `AffirmsReceived` and `OffChainAffirmations` storage.
+    fn ensure_no_missing_affirmation(
+        instruction_id: &InstructionId,
+        instruction_legs: &[(LegId, Leg)],
+    ) -> DispatchResult {
+        let mut unique_portfolios = BTreeSet::new();
+
+        for (leg_id, leg) in instruction_legs {
+            match leg {
+                Leg::Fungible {
+                    sender, receiver, ..
+                }
+                | Leg::NonFungible {
+                    sender, receiver, ..
+                } => {
+                    if unique_portfolios.insert(sender) {
+                        let sdr_affirmation_status = UserAffirmations::take(sender, instruction_id);
+                        ensure!(
+                            sdr_affirmation_status == AffirmationStatus::Affirmed,
+                            Error::<T>::NotAllAffirmationsHaveBeenReceived
+                        );
+                        let sdr_affirmation_status = AffirmsReceived::take(instruction_id, sender);
+                        ensure!(
+                            sdr_affirmation_status == AffirmationStatus::Affirmed,
+                            Error::<T>::NotAllAffirmationsHaveBeenReceived
+                        );
```

Comment:
I can't see any way for `UserAffirmations` and `AffirmsReceived` to be out of sync (i.e. for one to be Affirmed whilst the other is not) - are there any known ways for this to happen?

Reply:
It shouldn't, but we need to remove the old values anyways.  Those two values are simple/small, so it is cheap to decode and check the value.  We could just use `remove` for one and still use `take` for the other.